libbeat/monitoring/inputmon/input.go

```diff
@@ -19,23 +19,115 @@ package inputmon
 
 import (
 	"encoding/json"
+	"errors"
 	"strings"
+	"sync"
 
 	"github.com/gofrs/uuid/v5"
 
 	"github.com/elastic/elastic-agent-libs/logp"
 	"github.com/elastic/elastic-agent-libs/monitoring"
 )
 
-// NewInputRegistry returns a new monitoring.Registry for metrics related to
-// an input instance. The returned registry will be initialized with a static
-// string values for the input and id. When the input stops it should invoke
-// the returned cancel function to unregister the metrics. For testing purposes
-// an optional monitoring.Registry may be provided as an alternative to using
-// the global 'dataset' monitoring namespace. The inputType and id must be
-// non-empty for the metrics to be published to the global 'dataset' monitoring
-// namespace.
-func NewInputRegistry(inputType, id string, optionalParent *monitoring.Registry) (reg *monitoring.Registry, cancel func()) {
+type inputRegistry struct {
+	mu         sync.RWMutex
+	registries map[string]*monitoring.Registry
+}
+
+var registeredInputs = inputRegistry{
+	registries: make(map[string]*monitoring.Registry),
+}
+
+// RegisterMetrics adds reg to the collection of registries to be returned by
+// the `/inputs/` endpoint. The registry must have at least a `id` and `input`
+// string variables, otherwise the registry is rejected and an error is
+// returned.
+// If an id/inputType registry has benn already registered, it'll be overridden.
+// When the input finishes, it should call UnregisterMetrics to
+// release the associated resources.
+func RegisterMetrics(id string, reg *monitoring.Registry) error {
+	idValid := validStringVar(reg.Get("id"))
+	inputValid := validStringVar(reg.Get("input"))
+
+	errMgs := ""
+	if !idValid {
+		errMgs = "'id' empty or absent"
+	}
+	if !inputValid {
+		errMgs += ", 'input' empty or absent"
```

Comment:
When `id` is valid and input is not valid, this error message is going to be wonky.

And regarding `errMgs`, was that meant to be "Msg"?

Reply:
done in https://github.com/elastic/beats/pull/42618/commits/1ccbb086b534f884338d148d0a10687039d273b2